cmd/registry/cmd/upload/bulk/openapi.go

```diff
@@ -166,78 +165,82 @@ func (task *uploadOpenAPITask) populateFields() error {
 }
 
 func (task *uploadOpenAPITask) createAPI(ctx context.Context) error {
-	if _, err := task.client.GetApi(ctx, &rpc.GetApiRequest{
-		Name: task.apiName(),
-	}); !core.NotFound(err) {
-		return err // Returns nil when API is found without error.
-	}
-
-	response, err := task.client.CreateApi(ctx, &rpc.CreateApiRequest{
-		Parent: task.projectName() + "/locations/global",
-		ApiId:  task.apiID,
+	// Create an API if needed (or update an existing one)
+	response, err := task.client.UpdateApi(ctx, &rpc.UpdateApiRequest{
 		Api: &rpc.Api{
+			Name:        task.apiName(),
 			DisplayName: task.apiID,
 		},
+		AllowMissing: true,
 	})
 	if err == nil {
 		log.Debugf("Created %s", response.Name)
-	} else if core.AlreadyExists(err) {
-		log.Debugf("Found %s", task.apiName())
 	} else {
 		log.WithError(err).Debugf("Failed to create API %s", task.apiName())
+		// Returning this error ends all tasks, which seems appropriate to
+		// handle situations where all might fail due to a common problem
+		// (a missing project or incorrect project-id).
 		return fmt.Errorf("Failed to create %s, %s", task.apiName(), err)
 	}
 
 	return nil
 }
 
 func (task *uploadOpenAPITask) createVersion(ctx context.Context) error {
-	if _, err := task.client.GetApiVersion(ctx, &rpc.GetApiVersionRequest{
-		Name: task.versionName(),
-	}); !core.NotFound(err) {
-		return err // Returns nil when version is found without error.
-	}
-
-	response, err := task.client.CreateApiVersion(ctx, &rpc.CreateApiVersionRequest{
-		Parent:       task.apiName(),
-		ApiVersionId: task.versionID,
-		ApiVersion:   &rpc.ApiVersion{},
+	// Create an API version if needed (or update an existing one)
+	response, err := task.client.UpdateApiVersion(ctx, &rpc.UpdateApiVersionRequest{
+		ApiVersion: &rpc.ApiVersion{
+			Name: task.versionName(),
+		},
+		AllowMissing: true,
 	})
-	if err != nil {
-		log.WithError(err).Debugf("Failed to create version %s", task.versionName())
-	} else {
+	if err == nil {
 		log.Debugf("Created %s", response.Name)
+	} else {
+		log.WithError(err).Debugf("Failed to create version %s", task.versionName())
 	}
 
 	return nil
 }
 
-func (task *uploadOpenAPITask) createSpec(ctx context.Context) error {
-	contents, err := task.gzipContents()
+func (task *uploadOpenAPITask) createOrUpdateSpec(ctx context.Context) error {
+	contents, err := ioutil.ReadFile(task.path)
 	if err != nil {
 		return err
 	}
 
-	if _, err = task.client.GetApiSpec(ctx, &rpc.GetApiSpecRequest{
+	// Use the spec size and hash to avoid unnecessary uploads.
+	if spec, err := task.client.GetApiSpec(ctx, &rpc.GetApiSpecRequest{
 		Name: task.specName(),
-	}); !core.NotFound(err) {
-		return err // Returns nil when spec is found without error.
+	}); err == nil {
+		if int(spec.GetSizeBytes()) == len(contents) {
+			hash := hashForBytes(contents)
+			if spec.GetHash() == hash {
+				log.Debugf("Matched existing spec %s", task.specName())
+				return nil // this spec is already uploaded
+			}
+		}
+	}
```

Comment:
Does it make sense to refactor this into a helper? It seems identical to the discovery doc and proto functionality.

Reply:
Maybe, I noticed a lot of similarity in the three uploaders. I think that's worth watching.